From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: Besar Wicaksono <bwicaksono@nvidia.com>
Date: Fri, 3 Nov 2023 18:48:40 -0500
Subject: [PATCH 046/240] NVIDIA: SAUCE: perf: arm_cspmu: add pmpidr support

The PMIIDR value is composed by the values spread access
in PMPIDR registers. We can use PMPIDR registers as
alternative for device identification for systems that
do not implement PMIIDR.

Change-Id: I25ada0e24a7c773df6a21778fed9fc633f497c29
Signed-off-by: Besar Wicaksono <bwicaksono@nvidia.com>

diff --git a/drivers/perf/arm_cspmu/arm_cspmu.c b/drivers/perf/arm_cspmu/arm_cspmu.c
index e0d4293f06f93..00bdf6452512e 100644
--- a/drivers/perf/arm_cspmu/arm_cspmu.c
+++ b/drivers/perf/arm_cspmu/arm_cspmu.c
@@ -40,6 +40,75 @@
 	ARM_CSPMU_EXT_ATTR(_name, arm_cspmu_cpumask_show,	\
 				(unsigned long)_config)
 
+/*
+ * CoreSight PMU Arch register offsets.
+ */
+#define PMEVCNTR_LO					0x0
+#define PMEVCNTR_HI					0x4
+#define PMEVTYPER					0x400
+#define PMCCFILTR					0x47C
+#define PMEVFILTR					0xA00
+#define PMCNTENSET					0xC00
+#define PMCNTENCLR					0xC20
+#define PMINTENSET					0xC40
+#define PMINTENCLR					0xC60
+#define PMOVSCLR					0xC80
+#define PMOVSSET					0xCC0
+#define PMCFGR						0xE00
+#define PMCR						0xE04
+#define PMIIDR						0xE08
+#define PMPIDR0						0xFE0
+#define PMPIDR1						0xFE4
+#define PMPIDR2						0xFE8
+#define PMPIDR3						0xFEC
+#define PMPIDR4						0xFD0
+
+/* PMCFGR register field */
+#define PMCFGR_NCG					GENMASK(31, 28)
+#define PMCFGR_HDBG					BIT(24)
+#define PMCFGR_TRO					BIT(23)
+#define PMCFGR_SS					BIT(22)
+#define PMCFGR_FZO					BIT(21)
+#define PMCFGR_MSI					BIT(20)
+#define PMCFGR_UEN					BIT(19)
+#define PMCFGR_NA					BIT(17)
+#define PMCFGR_EX					BIT(16)
+#define PMCFGR_CCD					BIT(15)
+#define PMCFGR_CC					BIT(14)
+#define PMCFGR_SIZE					GENMASK(13, 8)
+#define PMCFGR_N					GENMASK(7, 0)
+
+/* PMCR register field */
+#define PMCR_TRO					BIT(11)
+#define PMCR_HDBG					BIT(10)
+#define PMCR_FZO					BIT(9)
+#define PMCR_NA						BIT(8)
+#define PMCR_DP						BIT(5)
+#define PMCR_X						BIT(4)
+#define PMCR_D						BIT(3)
+#define PMCR_C						BIT(2)
+#define PMCR_P						BIT(1)
+#define PMCR_E						BIT(0)
+
+/* PMPIDR0 register field */
+#define PMPIDR0_PART_0					GENMASK(7, 0)
+
+/* PMPIDR1 register field */
+#define PMPIDR1_DES_0					GENMASK(7, 4)
+#define PMPIDR1_PART_1					GENMASK(3, 0)
+
+/* PMPIDR2 register field */
+#define PMPIDR2_REVISION				GENMASK(7, 4)
+#define PMPIDR2_DES_1					GENMASK(2, 0)
+
+/* PMPIDR3 register field */
+#define PMPIDR3_REVAND					GENMASK(7, 4)
+#define PMPIDR3_CMOD					GENMASK(3, 0)
+
+/* PMPIDR4 register field */
+#define PMPIDR4_SIZE					GENMASK(7, 4)
+#define PMPIDR4_DES_2					GENMASK(3, 0)
+
 /* Each SET/CLR register supports up to 32 counters. */
 #define ARM_CSPMU_SET_CLR_COUNTER_SHIFT		5
 #define ARM_CSPMU_SET_CLR_COUNTER_NUM		\
@@ -351,6 +420,44 @@ static struct arm_cspmu_impl_match *arm_cspmu_impl_match_get(u32 pmiidr)
 	return NULL;
 }
 
+static u32 arm_cspmu_get_pmiidr(struct arm_cspmu *cspmu)
+{
+	u32 pmiidr, pmpidr;
+
+	pmiidr = readl(cspmu->base0 + PMIIDR);
+
+	if (pmiidr != 0)
+		return pmiidr;
+
+	/* Construct PMIIDR value from PMPIDRs. */
+
+	pmpidr = readl(cspmu->base0 + PMPIDR0);
+	pmiidr |= FIELD_PREP(ARM_CSPMU_PMIIDR_PRODUCTID_PART_0,
+				FIELD_GET(PMPIDR0_PART_0, pmpidr));
+
+	pmpidr = readl(cspmu->base0 + PMPIDR1);
+	pmiidr |= FIELD_PREP(ARM_CSPMU_PMIIDR_PRODUCTID_PART_1,
+				FIELD_GET(PMPIDR1_PART_1, pmpidr));
+	pmiidr |= FIELD_PREP(ARM_CSPMU_PMIIDR_IMPLEMENTER_DES_0,
+				FIELD_GET(PMPIDR1_DES_0, pmpidr));
+
+	pmpidr = readl(cspmu->base0 + PMPIDR2);
+	pmiidr |= FIELD_PREP(ARM_CSPMU_PMIIDR_VARIANT,
+				FIELD_GET(PMPIDR2_REVISION, pmpidr));
+	pmiidr |= FIELD_PREP(ARM_CSPMU_PMIIDR_IMPLEMENTER_DES_1,
+				FIELD_GET(PMPIDR2_DES_1, pmpidr));
+
+	pmpidr = readl(cspmu->base0 + PMPIDR3);
+	pmiidr |= FIELD_PREP(ARM_CSPMU_PMIIDR_REVISION,
+				FIELD_GET(PMPIDR3_REVAND, pmpidr));
+
+	pmpidr = readl(cspmu->base0 + PMPIDR4);
+	pmiidr |= FIELD_PREP(ARM_CSPMU_PMIIDR_IMPLEMENTER_DES_2,
+				FIELD_GET(PMPIDR4_DES_2, pmpidr));
+
+	return pmiidr;
+}
+
 #define DEFAULT_IMPL_OP(name)	.name = arm_cspmu_##name
 
 static int arm_cspmu_init_impl_ops(struct arm_cspmu *cspmu)
@@ -361,7 +468,7 @@ static int arm_cspmu_init_impl_ops(struct arm_cspmu *cspmu)
 
 	/* Start with a default PMU implementation */
 	cspmu->impl.module = THIS_MODULE;
-	cspmu->impl.pmiidr = readl(cspmu->base0 + PMIIDR);
+	cspmu->impl.pmiidr = arm_cspmu_get_pmiidr(cspmu);
 	cspmu->impl.ops = (struct arm_cspmu_impl_ops) {
 		DEFAULT_IMPL_OP(get_event_attrs),
 		DEFAULT_IMPL_OP(get_format_attrs),
diff --git a/drivers/perf/arm_cspmu/arm_cspmu.h b/drivers/perf/arm_cspmu/arm_cspmu.h
index 19684b76bd969..779dd2d6aa26b 100644
--- a/drivers/perf/arm_cspmu/arm_cspmu.h
+++ b/drivers/perf/arm_cspmu/arm_cspmu.h
@@ -115,8 +115,15 @@
 #define PMCR_E				BIT(0)
 
 /* PMIIDR register field */
-#define ARM_CSPMU_PMIIDR_IMPLEMENTER	GENMASK(11, 0)
-#define ARM_CSPMU_PMIIDR_PRODUCTID	GENMASK(31, 20)
+#define ARM_CSPMU_PMIIDR_IMPLEMENTER		GENMASK(11, 0)
+#define ARM_CSPMU_PMIIDR_IMPLEMENTER_DES_0	GENMASK(3, 0)
+#define ARM_CSPMU_PMIIDR_IMPLEMENTER_DES_1	GENMASK(6, 4)
+#define ARM_CSPMU_PMIIDR_IMPLEMENTER_DES_2	GENMASK(11, 8)
+#define ARM_CSPMU_PMIIDR_REVISION		GENMASK(15, 12)
+#define ARM_CSPMU_PMIIDR_VARIANT		GENMASK(19, 16)
+#define ARM_CSPMU_PMIIDR_PRODUCTID		GENMASK(31, 20)
+#define ARM_CSPMU_PMIIDR_PRODUCTID_PART_0	GENMASK(27, 20)
+#define ARM_CSPMU_PMIIDR_PRODUCTID_PART_1	GENMASK(31, 28)
 
 /* JEDEC-assigned JEP106 identification code */
 #define ARM_CSPMU_IMPL_ID_NVIDIA	0x36B
-- 
2.43.0

